spark/src/main/java/org/apache/iceberg/actions/RewriteDataFilesAction.java

```diff
@@ -0,0 +1,281 @@
+import org.apache.iceberg.DataFile;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.PartitionSpec;
+import org.apache.iceberg.RewriteFiles;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.TableProperties;
+import org.apache.iceberg.encryption.EncryptionManager;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.io.CloseableIterable;
+import org.apache.iceberg.io.CloseableIterator;
+import org.apache.iceberg.io.FileIO;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+import org.apache.iceberg.relocated.com.google.common.collect.ListMultimap;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+import org.apache.iceberg.relocated.com.google.common.collect.Multimaps;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.apache.iceberg.relocated.com.google.common.collect.Streams;
+import org.apache.iceberg.spark.source.RowDataRewriter;
+import org.apache.iceberg.util.PropertyUtil;
+import org.apache.iceberg.util.StructLikeWrapper;
+import org.apache.iceberg.util.TableScanUtil;
+import org.apache.iceberg.util.Tasks;
+import org.apache.spark.api.java.JavaRDD;
+import org.apache.spark.api.java.JavaSparkContext;
+import org.apache.spark.broadcast.Broadcast;
+import org.apache.spark.sql.SparkSession;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class RewriteDataFilesAction
+    extends BaseSnapshotUpdateAction<RewriteDataFilesAction, RewriteDataFilesActionResult> {
+
+  private static final Logger LOG = LoggerFactory.getLogger(RewriteDataFilesAction.class);
+
+  private final JavaSparkContext sparkContext;
+  private final Table table;
+  private final FileIO fileIO;
+  private final EncryptionManager encryptionManager;
+  private final boolean caseSensitive;
+  private long targetSizeInBytes;
+  private int splitLookback;
+  private long splitOpenFileCost;
+
+  private PartitionSpec spec = null;
+  private Expression filter;
+
+  RewriteDataFilesAction(SparkSession spark, Table table) {
+    this.sparkContext = new JavaSparkContext(spark.sparkContext());
+    this.table = table;
+    this.spec = table.spec();
+    this.filter = Expressions.alwaysTrue();
+    this.caseSensitive = Boolean.parseBoolean(spark.conf().get("spark.sql.caseSensitive", "false"));
+
+    long splitSize = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.SPLIT_SIZE,
+        TableProperties.SPLIT_SIZE_DEFAULT);
+    long targetFileSize = PropertyUtil.propertyAsLong(
+        table.properties(),
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES,
+        TableProperties.WRITE_TARGET_FILE_SIZE_BYTES_DEFAULT);
+    this.targetSizeInBytes = Math.min(splitSize, targetFileSize);
```

Comment:
Hm, if my split size is 128 MB and target file size is 512 MB, what is the size of files that we will create?

Reply:
Hm. Split size is how much data per task while reading, and write target size is how much data to put in a file when writing. The behavior of the write target size is more of an upper bound -- files could be smaller and Iceberg has no ability to coalesce to larger sizes. I think this should ignore the target file size because that's enforced in the writer regardless. Only the split size should matter.

Reply:
I'd consider packing files into bins that are around `targetFileSize` as defined by the table property and ignoring the threshold during writes by setting `target-file-size-bytes` to the max long value so that we avoid closing files near the end of bins. That way, we will get files that are close to the target file size.

Reply:
After thinking about this a bit more, I think I wasn't correct that target write size is a maximum. It's currently used that way in Spark, but the property is "target" and that implies that it is the size files _should_ be, ideally. So Anton is right and that's the size of files we should attempt to create. This also makes sense because it is a `write` property, not a `read` property.

I think we should remove the min with split size and just use the table's target size as the default.

Reply:
Let's also set `target-file-size-bytes` to the max long value while writing so that we avoid closing files near the end of bins. For example, if we ended up with 520 MB of data in a task, it is better to write it as one file instead of 512MB and 8 MB.

Reply:
I am also OK addressing this in a follow-up as the PR itself is large.

Reply:
@aokolnychyi @rdblue , let me rephrase the purpose you mentioned above:

1. Using the `targetFileSize` to do the split, and also the written file size.
2. Setting `write.target-file-size-bytes` to `Long.MAX_VALUE` to make sure there's no separated files for one written task.

Am I understanding right? 

Besides, what is the default value of `targetFileSize`, should it be explicitly defined by user, or use `SPLIT_SIZE` as default?

Reply:
I think the conclusion was that the planning in this feature should use `write.target-file-size-bytes` for the default `targetFileSize` when planning the splits because we should use the target write size (instead of the target read/split size for most reads) for our task size.

Next, we also need to override `write.target-file-size-bytes` for the write itself because we expect the write to already produce files close to that size, and leaving it at the normal target is likely to produce tiny files when a task's output is just over the limit.

Reply:
Thanks @rdblue ,

The default value of `write.target-file-size-bytes` is `Long.MAX_VALUE`. If users don't override this configuration explicitly, and use this to do planning ( bin-packing), will this lead to the problem where all the `FileScanTask`s combined into one?